packages/react-reconciler/src/ReactFiberCommitWork.old.js

```diff
@@ -2414,24 +2416,54 @@ function getRetryCache(finishedWork) {
 }
 
 export function detachOffscreenInstance(instance: OffscreenInstance): void {
-  const currentOffscreenFiber = instance._current;
-  if (currentOffscreenFiber === null) {
+  const fiber = instance._current;
+  if (fiber === null) {
     throw new Error(
       'Calling Offscreen.detach before instance handle has been set.',
     );
   }
 
-  const executionContext = getExecutionContext();
-  if ((executionContext & (RenderContext | CommitContext)) !== NoContext) {
-    scheduleMicrotask(() => {
-      instance._visibility |= OffscreenDetached;
-      disappearLayoutEffects(currentOffscreenFiber);
-      disconnectPassiveEffect(currentOffscreenFiber);
-    });
-  } else {
+  if ((instance._visibility & OffscreenDetached) !== NoFlags) {
+    // The instance is already detached, this is a noop.
+    return;
+  }
+
+  instance._pendingVisibility |= OffscreenDetached;
+
+  // Detaching needs to be postoned in case attach is called before next update.
+  scheduleMicrotask(() => {
+    if ((instance._pendingVisibility & OffscreenDetached) === NoFlags) {
+      // Attach was called. Offscreen does not need to be detached.
+      return;
+    }
+
     instance._visibility |= OffscreenDetached;
```

Comment:
This doesn't look right to me. `_pendingVisibility` is represents the queued state. `_visibility` represents the current state. So it shouldn't be set until the commit phase.

Reply:
I changed it so `attachOffscreenInstance` and `detachOffscreenInstance` only read/write to `_pendingVisibility` and `_pendingVisibility` is synced with `_visibility` in `beginWork`.